subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/ivyresolve/CachingModuleComponentRepository.java

```diff
@@ -335,19 +338,30 @@ private void resolveArtifactFromCache(ComponentArtifactMetadata artifact, Cachin
     }
 
     private class ResolveAndCacheRepositoryAccess implements ModuleComponentRepositoryAccess {
+        private final CacheLockingManager cacheLockingManager;
+
+        ResolveAndCacheRepositoryAccess(CacheLockingManager cacheLockingManager) {
+            this.cacheLockingManager = cacheLockingManager;
+        }
+
         @Override
         public String toString() {
             return "cache > " + delegate.getRemoteAccess().toString();
         }
 
         @Override
-        public void listModuleVersions(DependencyMetadata dependency, BuildableModuleVersionListingResolveResult result) {
+        public void listModuleVersions(final DependencyMetadata dependency, final BuildableModuleVersionListingResolveResult result) {
             delegate.getRemoteAccess().listModuleVersions(dependency, result);
             switch (result.getState()) {
                 case Listed:
-                    ModuleIdentifier moduleId = getCacheKey(dependency.getRequested());
-                    Set<String> versionList = result.getVersions();
-                    moduleVersionsCache.cacheModuleVersionList(delegate, moduleId, versionList);
+                    cacheLockingManager.useCache(new Runnable() {
```

Comment:
That's what I was looking for! I couldn't remember what you guys had called that.